plugins/outputs/tcp_forwarder/tcp_forwarder.go

```diff
@@ -0,0 +1,148 @@
+  servers = ["localhost:8089"]
+  ## timeout in seconds for the write connection
+  timeout = 2
+  ## reconnect before every push
+  reconnect = false
+  ## Data format to consume.
+  ## Each data format has it's own unique set of configuration options, read
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_INPUT.md
+  data_format = "influx"
+`
+
+// SetSerializer is the function from output plugin to use a serializer for
+// formating data
+func (t *TCPForwarder) SetSerializer(serializer serializers.Serializer) {
+	t.serializer = serializer
+}
+
+// Connect is the default output plugin connection function who make sure it
+// can connect to the endpoint
+func (t *TCPForwarder) Connect() error {
+
+	if t.Timeout <= 0 {
+	}
+	if len(t.Servers) == 0 {
+		t.Servers = append(t.Servers, "localhost:8089")
+	}
+
+	// try connect
+	t.conns = make(map[string]net.Conn, len(t.Servers))
+	t.reconnect()
+	return nil
+}
+
+func (t *TCPForwarder) reconnect() error {
+
+	// Create a map of connection with endpoint as key and connection as value
+	for _, endpoint := range t.Servers {
+		if value, ok := t.conns[endpoint]; !ok || value == nil {
+			fmt.Println("reconnect: " + endpoint)
+			conn, err := net.DialTimeout("tcp", endpoint, time.Duration(t.Timeout)*time.Second)
+			if err == nil {
+				t.conns[endpoint] = conn
+			} else {
+				log.Printf("Error connecting to <%s>: %s", endpoint, err.Error())
+			}
+		}
+	}
+	return nil
+}
+
+// Close is use to close connection to all Tcp endpoints
+func (t *TCPForwarder) Close() error {
+	// Closing all connections
+	for endpoint := range t.conns {
+		t.conns[endpoint].Close()
+		delete(t.conns, endpoint)
+	}
+	return nil
+}
+
+// SampleConfig is the default function who return the default configuration
+// for tcp forwarder output
+func (t *TCPForwarder) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description is the default function who return the description of tcp
+// forwarder output
+func (t *TCPForwarder) Description() string {
+	return "Generic TCP forwarder for metrics"
+}
+
+// Write is the default function to call to "send" a metric through the Output
+// Choose a random endpoint in the cluster to write to until a successful write
+// occurs, logging each unsuccessful. If all servers fail, return error.
+func (t *TCPForwarder) Write(metrics []telegraf.Metric) error {
+	// reconnect if needed
+	t.reconnect()
+	// Prepare data
+	var bp []string
+	for _, metric := range metrics {
+		sMetrics, err := t.serializer.Serialize(metric)
+		if err != nil {
+			log.Printf("Error while serializing some metrics: %s", err.Error())
+		}
+		bp = append(bp, sMetrics...)
+	}
+	// TODO should we add a join function in serialiser ?
+	points := strings.Join(bp, "\n") + "\n"
+
+	// This will get set to nil if a successful write occurs
+	err := errors.New("Could not write to any tcp endpoint\n")
+
+	// Send data to a random endpoint
+	p := rand.Perm(len(t.Servers))
```

Comment:
you have not documented that this output selects a random server to write to, thereby load-balancing the output servers defined.

Was this done on purpose? I'm not sure it should be, since it's a generic TCP output. Where we have done this in the past is only for output sinks that are clustered, and thus only need one IP address to be written to.